app/controllers/EpicTestsController.scala

```diff
@@ -76,14 +89,17 @@ class EpicTestsController(authAction: AuthAction[AnyContent], components: Contro
   def archivedTestNames = authAction.async { request =>
     val objectSettings = archiveObjectSettings(stage, fileName = "")
 
-    s3Client.listKeys(objectSettings).map {
-      case Right(keys) =>
-        val testNames: List[String] = keys.flatMap(extractTestName)
-        Ok(S3Json.noNulls(testNames.asJson))
-
-      case Left(error) =>
-        logger.error(s"Failed to fetch list of archived test names: $error")
-        InternalServerError(error)
+    run {
+      s3Client
+        .listKeys(objectSettings)
+        .map { keys =>
+          val testNames: List[String] = keys.flatMap(extractTestName)
+          Ok(S3Json.noNulls(testNames.asJson))
+        }
+        .mapError { error =>
```

Comment:
Use [`tapError`](https://javadoc.io/static/dev.zio/zio_2.12/1.0.0-RC17/zio/ZIO.html#tapError[R1<:R,E1>:E](f:E=>zio.ZIO[R1,E1,Any])(implicitev:zio.CanFail[E]):zio.ZIO[R1,E1,A]) instead.

Reply:
(logging is an effect which can throw)

Reply:
What does `tapError` do if the logging throws? Provide the original error, or the one produced from the logging?

Reply:
Errors are never lost in ZIO. They are accumulated in sequence or in parallel, depending. An action produces an `Exit[E, A]` which can either be a success, or a failure. In the latter case, the failure contains a [`Cause[E]`](https://javadoc.io/doc/dev.zio/zio_2.12/latest/zio/Cause.html) where you have access to all the errors that happened.

Reply:
Thanks @regiskuckaertz, I really want to have play with zio, it looks like a breath of fresh air.

Reply:
In this case `tapError` works fine. In cases where I want to narrow the type of the error it fails because of the type bound on the error type in `tapError[R1 <: R, E1 >: E](f: E => ZIO[R1, E1, _]): ZIO[R1, E1, A]`
E.g.
```
[error]  found   : zio.ZIO[zio.blocking.Blocking,Throwable,services.VersionedS3Data[String]]
[error]  required: zio.ZIO[zio.blocking.Blocking,services.S3Client.S3ClientError,services.S3Client.RawVersionedS3Data]
```
I think I just have to stick with `mapError` in these cases?

Reply:
If your code looks sth like this:
```scala
.tapError(x => Task(logger.error(...)))
```
Then it makes sense, as `Throwable` is the least upper bound between `S3ClientError` and `Throwable`. Since logging is not essential, I suggest you just forget the error, like this:
```scala
.tapError(x => Task(logger.error(...)).ignore)
```